services.go

```diff
@@ -91,7 +92,7 @@ func (s *serviceSet) dispatch(ctx context.Context, serviceName, methodName strin
 	}
 
 	resp, err := s.interceptor(ctx, unmarshal, info, method)
-	if err != nil {
+	if err != nil || isNil(resp) {
```

Comment:
I think it is not good approach to return nil here. What about return the error `error while marshaling`?
ref: https://github.com/grpc/grpc-go/pull/1251/files

Reply:
Since marshaling on nil will crash directly, and here there is no chance to get  a returned value.  Actually I had tried grpc, and the same issue also existed on it.

Reply:
I changed the example from grpc-go. It returns internal error without crash.

```
root@ubuntu-xenial ~/g/s/g/g/e/h/greeter_server# go run main.go                                                                                                                                                                                                       1 master!
2019/10/16 18:53:16 Received: world
ERROR: 2019/10/16 18:53:16 grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: proto: Marshal called with nil
```

My point is that the value should be empty not nil. Otherwise, you can to check twice for the return. One is for error, and other one is for value. If the resp is nil, I think it is fine to return error like marshal error without crash.

Reply:
Maybe a typed error message when there is no response and we can switch on that as a better option?  I don't like functions that `return nil, nil` as they just ask for trouble and I feel like this is a coding mistake

Reply:
Hi @crosbymichael and @fuweid , yes, It's really a coding mistake, and 

> I changed the example from grpc-go. It returns internal error without crash.
> 
> ```
> root@ubuntu-xenial ~/g/s/g/g/e/h/greeter_server# go run main.go                                                                                                                                                                                                       1 master!
> 2019/10/16 18:53:16 Received: world
> ERROR: 2019/10/16 18:53:16 grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: proto: Marshal called with nil
> ```

Hi @fuweid, what's the  type you returned in your test codes?  I returned a nil of proto.Empty, and it did crashed the grpc server. 
Yes, protobuf defined an error type of 'ErrNil' at https://github.com/gogo/protobuf/blob/master/proto/encode.go#L53, but it seems not all of the messages type's Marshal() do the nil check, I think that's why some type's nil would trigger 'ErrNil' returned but others would crash the server.

Thus I think's it's better to do a nil check before doing message's Marshal.

Reply:
> Maybe a typed error message when there is no response and we can switch on that as a better option? I don't like functions that `return nil, nil` as they just ask for trouble and I feel like this is a coding mistake

Hi @crosbymichael 
I agree.  How about defining an error type 'ErrNil'  just as protobuf did and 'return nil, ErrNil'   instead of  ' return nil, nil' ?